lib/registry.py

```diff
@@ -0,0 +1,584 @@
+      ch.INFO("manifest: uploading")
+      url = self._url_of("manifests", self.ref.tag)
+      self.request("PUT", url, {201}, data=manifest,
+                   headers={ "Content-Type": TYPES_MANIFEST["docker2"] })
+
+   def request(self, method, url, statuses={200}, out=None, **kwargs):
+      """Request url using method and return the response object. If statuses
+         is given, it is set of acceptable response status codes, defaulting
+         to {200}; any other response is a fatal error. If out is given,
+         response content will be streamed to this Progress_Writer object and
+         must be non-zero length.
+
+         Use current session if there is one, or start a new one if not. If
+         authentication fails (or isn't initialized), then authenticate harder
+         and re-try the request."""
+      # Set up.
+      self.session_init_maybe()
+      ch.VERBOSE("auth: %s" % self.auth)
+      if (out is not None):
+         kwargs["stream"] = True
+      # Make the request.
+      while True:
+         res = self.request_raw(method, url, statuses | {401}, **kwargs)
+         if (res.status_code != 401):
+            break
+         else:
+            ch.VERBOSE("HTTP 401 unauthorized")
+            if (self.escalate(res)):   # success
+               ch.VERBOSE("retrying with auth: %s" % self.auth)
+            elif (401 in statuses):    # caller can deal with it
+               break
+            else:
+               ch.FATAL("unhandled authentication failure")
+      # Stream response if needed.
+      if (out is not None and res.status_code == 200):
+         try:
+            length = int(res.headers["Content-Length"])
+         except KeyError:
+            length = None
+         except ValueError:
+            ch.FATAL("invalid Content-Length in response")
+         out.start(length)
+         for chunk in res.iter_content(ch.HTTP_CHUNK_SIZE):
+            out.write(chunk)
+      # Done.
+      return res
+
+   def request_raw(self, method, url, statuses, auth=None, **kwargs):
+      """Request url using method. statuses is an iterable of acceptable
+         response status codes; any other response is a fatal error. Return
+         the requests.Response object.
+
+         Session must already exist. If auth arg given, use it; otherwise, use
+         object's stored authentication if initialized; otherwise, use no
+         authentication."""
+      ch.VERBOSE("%s: %s" % (method, url))
+      if (auth is None):
+         auth = self.auth
+      try:
+         res = self.session.request(method, url, auth=auth, **kwargs)
+         ch.VERBOSE("response status: %d" % res.status_code)
+         if (res.status_code not in statuses):
+            ch.FATAL("%s failed; expected status %s but got %d: %s"
+                  % (method, statuses, res.status_code, res.reason))
+      except requests.exceptions.RequestException as x:
+         ch.FATAL("%s failed: %s" % (method, x))
+      self.headers_log(res.headers)
```

Comment:
This is my mistake, but we need to log the headers before we bail out on error. Move to before line 571?